src/main/java/org/testng/asserts/SoftAssert.java

```diff
@@ -43,6 +44,8 @@ public void assertAll() {
           sb.append(" ").append(cause.getMessage());
           cause = cause.getCause();
         }
+        
+        sb.append(StackTraceTools.formatByStackTrace(error.getStackTrace()[4]));
```

Comment:
How are we arriving at this number of `4` ?

Reply:
Hmm, this number is dependent on the calling order in the stack trace,
which now that you mention it could be fragile if there's a change in the calling order.

Do you think it would be worth making more dynamic? As in,
choosing that index by looking through the stack trace for the most
recent call of softAssert.equals/softAssert.fail? What are your thoughts on that :)

Reply:
What I meant was something like this

1. Add a new method into `Assertion` class which looks like this

```java
  protected String getErrorDetails(Throwable error) {
    StringBuilder sb = new StringBuilder();
    sb.append(error.getMessage());
    Throwable cause = error.getCause();
    while (cause != null) {
      sb.append(" ").append(cause.getMessage());
      cause = cause.getCause();
    }
    return sb.toString();
  }
```

Now change `assertAll()` method in `SoftAssert` to look like this

```java
  public void assertAll() {
    if (!m_errors.isEmpty()) {
      StringBuilder sb = new StringBuilder("The following asserts failed:");
      boolean first = true;
      for (AssertionError error : m_errors.keySet()) {
        if (first) {
          first = false;
        } else {
          sb.append(",");
        }
        sb.append("\n\t");
        sb.append(getErrorDetails(error));
      }
      throw new AssertionError(sb.toString());
    }
  }
```

This way you have enabled downstream consumers of `SoftAssert` to decide how they want to report errors . So in your case, you can subclass `SoftAssert` and plugin the logic of extracting out the first line in your test project and TestNG doesn't need to change at all.

That would also do away with having to guess the stack trace by TestNG, but is something that can be handled specifically within a test project that consumes TestNG.

Reply:
I really liked this idea, I added that into my PR.